experiments/compositions/composition/tests/testclient/client.go

```diff
@@ -219,6 +238,118 @@ func (c *Client) MustExist(objs []*unstructured.Unstructured, timeout time.Durat
+	c.T.Helper()
+
+	toMatch := ExtractGVKNN(obj)
+	c.T.Logf("Checking condition %q present in object %q on cluster %q", condition.Type, toMatch, c)
+
+	unmatched := []*unstructured.Unstructured{obj}
+	errMissing := fmt.Errorf("condition missing")
+	errMismatch := fmt.Errorf("condition mismatch")
+
+	retryFrequency := getFrequency(c.T, opDuration, timeout)
+	err := retry.OnError(retryFrequency, func(err error) bool {
+		return apierrors.IsNotFound(err) || errors.Is(err, errMissing) || errors.Is(err, errMismatch)
+	}, func() (err error) {
+		match := func(obj *unstructured.Unstructured) error {
+			read, err := c.Read(obj)
+			if err != nil {
+				return err
+			}
+			conditions, ok, err := unstructured.NestedSlice(read.Object, "status", "conditions")
+			if err != nil {
+				return err
+			}
+			if !ok {
+				return fmt.Errorf("%w: .status.conditions not found", errMissing)
+			}
+			found := false
+			for i := range conditions {
+				cond, ok := conditions[i].(map[string]interface{})
+				if !ok {
+					return fmt.Errorf("%w: .condition[%d] not of type map[string]interface{}", errMissing, i)
+				}
+				if condition.Type != cond["type"] {
+					continue
+				}
+				found = true
+				if condition.Reason != "" && condition.Reason != cond["reason"] {
+					return fmt.Errorf("%w: .condition.reason=%s not expected value %s", errMismatch, condition.Reason, cond["reason"])
+				}
+				if condition.Message != "" && condition.Message != cond["message"] {
+					return fmt.Errorf("%w: .condition.message=%s not expected value %s", errMismatch, condition.Message, cond["message"])
+				}
+				c.T.Logf("Has condition [%s, %s, %s]", cond["type"], cond["reason"], cond["message"])
+			}
+			if !found {
+				return fmt.Errorf("%w: .condition.type=%s not found", errMissing, condition.Type)
+			}
+			return nil
+		}
+		unmatched, err = c.recordFailed(match, unmatched)
+		return err
+	})
+	if err != nil {
+		c.T.Errorf("unexpected error: %v", err)
+		c.T.FailNow()
+	}
+}
+
+// MustNotHaveCondition - validate if the .status.conditions[] does not have the condition
+func (c *Client) MustNotHaveCondition(obj *unstructured.Unstructured, condition *metav1.Condition, timeout time.Duration) {
+	c.T.Helper()
+
+	toMatch := ExtractGVKNN(obj)
+	c.T.Logf("Checking condition %q not present in object %q on cluster %q", condition.Type, toMatch, c)
+
+	matched := []*unstructured.Unstructured{obj}
+	errMatches := fmt.Errorf("condition found")
+
+	retryFrequency := getFrequency(c.T, opDuration, timeout)
+	err := retry.OnError(retryFrequency, func(err error) bool {
+		return apierrors.IsNotFound(err) || errors.Is(err, errMatches)
+	}, func() (err error) {
+		match := func(obj *unstructured.Unstructured) error {
+			read, err := c.Read(obj)
+			if err != nil {
+				return err
+			}
+			conditions, ok, err := unstructured.NestedSlice(read.Object, "status", "conditions")
```

Comment:
BTW, when we're just reading, I really like defining my own status.conditions type and then using the runtime.DefaultUnstructuredConverter